controllers/securitygroups_controller.go

```diff
@@ -0,0 +1,286 @@
+	scope.Logger().Info("Reconciling SecurityGroups delete")
+
+	// Wait for machines to be deleted before removing the finalizer as they
+	// depend on this resource to deprovision.  Additionally it appears that
+	// allowing the Kubernetes API to vanish too quickly will upset the capi
+	// kubeadm control plane controller.
+	machines, err := collections.GetFilteredMachinesForCluster(ctx, r.Client, cluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(machines) != 0 {
+		scope.Logger().Info("Waiting for machines to be deleted", "remaining", len(machines))
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
+
+	if err = networkingService.DeleteBastionSecurityGroup(openStackCluster, fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)); err != nil {
+		handleUpdateOSCError(openStackCluster, fmt.Errorf("failed to delete bastion security group: %w", err))
+		return reconcile.Result{}, fmt.Errorf("failed to delete bastion security group: %w", err)
+	}
+	openStackCluster.Status.BastionSecurityGroup = nil
+
+	if err = networkingService.DeleteControlPlaneSecurityGroup(openStackCluster, clusterName); err != nil {
+		handleUpdateOSCError(openStackCluster, fmt.Errorf("failed to delete control plane security group: %w", err))
+		return reconcile.Result{}, fmt.Errorf("failed to delete control plane security group: %w", err)
+	}
+	openStackCluster.Status.ControlPlaneSecurityGroup = nil
+
+	if err = networkingService.DeleteWorkerSecurityGroup(openStackCluster, clusterName); err != nil {
+		handleUpdateOSCError(openStackCluster, fmt.Errorf("failed to delete worker security group: %w", err))
+		return reconcile.Result{}, fmt.Errorf("failed to delete worker security group: %w", err)
+	}
+	openStackCluster.Status.WorkerSecurityGroup = nil
+
+	scope.Logger().Info("Reconciled SecurityGroups deleted successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *SecurityGroupsReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(options).
+		For(
+			&infrav1.OpenStackMachine{},
+			builder.WithPredicates(
+				predicate.Funcs{
+					// Avoid reconciling if the event triggering the reconciliation is related to incremental status updates
+					UpdateFunc: func(e event.UpdateEvent) bool {
+						oldMachine := e.ObjectOld.(*infrav1.OpenStackMachine).DeepCopy()
+						newMachine := e.ObjectNew.(*infrav1.OpenStackMachine).DeepCopy()
+						oldMachine.Status = infrav1.OpenStackMachineStatus{}
+						newMachine.Status = infrav1.OpenStackMachineStatus{}
+						oldMachine.ObjectMeta.ResourceVersion = ""
+						newMachine.ObjectMeta.ResourceVersion = ""
+						return !reflect.DeepEqual(oldMachine, newMachine)
+					},
```

Comment:
I wouldn't bother adding something like this until you're sure you need it. In this case, I suspect this could be broken anyway.